src/phase0/state_transition/mod.rs

```diff
@@ -0,0 +1,387 @@
+    EPOCHS_PER_HISTORICAL_VECTOR,
+    EPOCHS_PER_SLASHINGS_VECTOR,
+    MAX_VALIDATORS_PER_COMMITTEE,
+    PENDING_ATTESTATIONS_BOUND,
+> {
+    state
+}
+
+pub fn verify_block_signature<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+    const MAX_PROPOSER_SLASHINGS: usize,
+    const MAX_ATTESTER_SLASHINGS: usize,
+    const MAX_ATTESTATIONS: usize,
+    const MAX_DEPOSITS: usize,
+    const MAX_VOLUNTARY_EXITS: usize,
+>(
+    state: BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    signed_block: SignedBeaconBlock<
+        MAX_PROPOSER_SLASHINGS,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        MAX_ATTESTER_SLASHINGS,
+        MAX_ATTESTATIONS,
+        MAX_DEPOSITS,
+        MAX_VOLUNTARY_EXITS,
+    >,
+) -> bool {
+    let proposer_index = signed_block.message.proposer_index;
+    let proposer = state
+        .validators
+        .get(proposer_index)
+        .expect("failed to get validator");
+    let domain = match get_domain(&state, DomainType::BeaconProposer, None) {
+        Ok(domain) => domain,
+        Err(_) => return false,
+    };
+    let signing_root = match compute_signing_root(&signed_block.message, domain) {
+        Ok(root) => root,
+        Err(_) => return false,
+    };
+
+    proposer
+        .pubkey
+        .verify_signature(signing_root, signed_block.signature)
+}
+
+pub fn get_domain<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    domain_type: DomainType,
+    epoch: Option<Epoch>,
+) -> Result<Domain, Error> {
+    let epoch = if epoch.is_none() {
+        get_current_epoch(&state)
+    } else {
+        epoch.unwrap()
+    };
+    let fork_version = if epoch < state.fork.epoch {
+        Some(state.fork.previous_version.clone())
+    } else {
+        Some(state.fork.current_version.clone())
+    };
```

Comment:
why did the `clone`s come back?

Reply:
because now it's the fastest way to prototype all primitives and then improve their API's.
If remove `clone()` we have to define all API first.